scheduler: avoid false positive detection of missed cluster event

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

When a cluster event for a pod arrives that normally would trigger scheduling of a pod while that pod is already being scheduled *and* that scheduling attempt fails with "Unschedulable", the priority queue put that pod into the backoff queue because it wouldn't be guaranteed that some other event would trigger the next scheduling attempt. In the backoff queue, the usual timeouts do that.

However, this "event observed" was tracked per queue, not per pod, so in a busy cluster with pods that typically fail at least once with "Unschedulable" (for example, pods using volumes or resources with delayed allocation), pods got moved to the backoff queue unnecessarily due to false positives: some event for a different pod was observed.

Those false positives get avoided completely when tracking per pod. A benchmark for scheduling pods with resource claim templates finishes in 60 seconds instead of 170s.

#### Special notes for your reviewer:

The downside is that some state only relevant for PriorityQueue gets placed in the shared QueuedPodInfo, which is not architecturally clean. A cleaner, but also more complicated solution would be to let queue implementations store their state in an interface{}. The pragmatic solution with int64 was chosen because it's more efficient (no need to cast, less memory allocations).

#### Does this PR introduce a user-facing change?
```release-note
scheduler: workloads that consists of many pods that need to be scheduled multiple times (for examples, pods with volumes and delayed binding or resource claims with delayed allocation) will get scheduled faster because pods get placed in the backoff queue less often (or not at all).
```

pohly committed:
scheduler: support events with custom "is schedulable" checks

Previously, plugins could only register events that then moved all pods blocked
by the plugin to the scheduling queue. This is too coarse for the dynamic
resource plugin: when a PodSchedulingContext object gets modified, that affects
exactly one Pod.

The most common sequence of events with dynamic resources is that the first
scheduling attempt collects information about potential nodes, creates a
PodSchedulingContext object, and then fails with Unschedulable status in
Reserve and waits for the resource driver to react by allocating
ResourceClaims. That ResourceClaim cluster event then put the Pod into the
backoff queue because it had been tried before. The backoff delay then slowed
down scheduling of the Pod.

To address both problems, a plugin may now provide an optional IsSchedulable
callback that checks how an informer event (add/update/delete) affects a
Pod. The result is not just binary (no/maybe) as before, but can also be a
stronger "schedule immediately" which puts the Pod into the normal scheduling
queue.

Tying the decision to back off or not to the event has the advantage that
multiple different events might make a new scheduling attempt necessary, but
not all of them necessarily need or should run immediately. The alternative
would have been a special "Waiting" status for the Pod, but then all events
would have led to skipping backoff.

Besides dynamic resource allocation, volume binding probably also could benefit
from such additional checks because it uses similar code patterns (trigger
volume creation during one scheduling cycle, then wait for it).

Without changes in plugins, performance is about the same as before. The next
commit changes the DRA plugin which then improves performance for single pod
scheduling (= latency).

pohly committed:
scheduler/dra: reduce pod scheduling latency

This is a combination of two related enhancements:
- By implementing a PreEnqueue check, the initial pod scheduling
  attempt for a pod with a claim template gets avoided when the claim
  does not exist yet.
- By implementing cluster event checks, only those pods get
  scheduled for which something changed, and they get scheduled
  immediately without delay.

This reduces the average latency for sequential scheduling of pods with
resource claim template from 5s to 0.02s (measured with an out-of-tree test
case in scheduler_perf).

pohly committed:
scheduler: more precise handling of cluster events

When a cluster event for a pod arrives that normally would trigger scheduling
of a pod while that pod is already being scheduled *and* that scheduling
attempt fails with "Unschedulable", the priority queue put that pod into the
backoff queue because it wouldn't be guaranteed that some other event would
trigger the next scheduling attempt. In the backoff queue, the usual timeouts
do that.

However, this "event observed" was tracked per queue, not per pod, so in a busy
cluster with pods that typically fail at least once with "Unschedulable" (for
example, pods using volumes or resources with delayed allocation), pods got
moved to the backoff queue unnecessarily due to false positives: some event for
a different pod was observed.

Those false positives get reduced when tracking per pod. A benchmark for
scheduling pods with resource claim templates finishes in 60 seconds instead of
170s.

The downside is that some state only relevant for PriorityQueue gets placed in
the shared QueuedPodInfo, which is not architecturally clean. A cleaner, but
also more complicated solution would be to let queue implementations store
their state in an interface{}. The pragmatic solution with int64 was chosen
because it's more efficient (no need to cast, less memory allocations).

This side effect of MoveAllToActiveOrBackoffQueue is important in several
different scenarios also for pods that are currently being scheduled without
being in queue of unschedulable pods. So to preserve that semantic, the queue
now keeps track of pods which are being processed.

pohly commented:
Let's look at one test failure, `TestPrebindPlugin/fail_on_1st_try_but_succeed_on_retry`. `AddUnschedulableIfNotPresent` gets called after a a PreBind failure. `pInfo.MoveRequestCycle` is zero, podSchedulingCycle is three, so the pod goes into the normal queue for unschedulable pods.

That's different from before. Previously, both `p.moveRequestCycle` and `podSchedulingCycle` were 1 and the pod went into the backoff queue.

So why is `p.moveRequestCycle == 1`? It got called here, before `AddUnschedulableIfNotPresent` is called:

```
E0216 09:18:10.779187   15825 framework.go:1063] "Failed running PreBind plugin" err="injecting failure for pod test-pod" plugin="prebind-plugin" pod="prebind-plugin60a2f08d-6ecf-4bcc-8297-583443e36358/test-pod" node="test-node-1"
> k8s.io/kubernetes/pkg/scheduler/internal/queue.(*PriorityQueue).movePodsToActiveOrBackoffQueue() ./pkg/scheduler/internal/queue/scheduling_queue.go:738 (hits goroutine(5301):1 total:1) (PC: 0x562b2cb)
   733:					metrics.SchedulerQueueIncomingPods.WithLabelValues("active", event.Label).Inc()
   734:					p.unschedulablePods.delete(pod, gated)
   735:				}
   736:			}
   737:		}
=> 738:		p.moveRequestCycle = p.schedulingCycle
   739:		if activated {
   740:			p.cond.Broadcast()
   741:		}
   742:	}
   743:	
(dlv) bt
0  0x000000000562b2cb in k8s.io/kubernetes/pkg/scheduler/internal/queue.(*PriorityQueue).movePodsToActiveOrBackoffQueue
   at ./pkg/scheduler/internal/queue/scheduling_queue.go:738
1  0x000000000562a725 in k8s.io/kubernetes/pkg/scheduler/internal/queue.(*PriorityQueue).MoveAllToActiveOrBackoffQueue
   at ./pkg/scheduler/internal/queue/scheduling_queue.go:705
2  0x00000000056c6dec in k8s.io/kubernetes/pkg/scheduler.(*Scheduler).handleBindingCycleError
   at ./pkg/scheduler/schedule_one.go:297
3  0x00000000056c5052 in k8s.io/kubernetes/pkg/scheduler.(*Scheduler).scheduleOne.func1
   at ./pkg/scheduler/schedule_one.go:113
4  0x00000000004784a1 in runtime.goexit
   at /nvme/gopath/go-1.19/src/runtime/asm_amd64.s:1594
(dlv) p activated
false
```

There was no pod that had to be activated. The only effect of `movePodsToActiveOrBackoffQueue` is that it sets `p.moveRequestCycle`. Looking at the failure handler, that seems to be the intent:
https://github.com/kubernetes/kubernetes/blob/6a5c88b9dd0a0e9ef0115d1bbcb196284777832b/pkg/scheduler/schedule_one.go#L286-L298

It carefully manages when `MoveAllToActiveOrBackoffQueue` gets called (with or without defer) to ensure that the following call to `FailureHandler` uses the "right" queue for the pod.

This worked, but it seems rather indirect: instead of simply telling `FailureHandler` that "this pod needs to go into the backoff queue" some global state gets changed to achieve that effect.

Note that this proposed change fails the same way:
```
	if activated {
		p.moveRequestCycle = p.schedulingCycle
		p.cond.Broadcast()
	}
```

This fixes this particular test failure:
```patch
diff --git a/pkg/scheduler/schedule_one.go b/pkg/scheduler/schedule_one.go
index cc0fbc73725..53f1f01bf53 100644
--- a/pkg/scheduler/schedule_one.go
+++ b/pkg/scheduler/schedule_one.go
@@ -287,13 +287,13 @@ func (sched *Scheduler) handleBindingCycleError(
                // as the assumed Pod had occupied a certain amount of resources in scheduler cache.
                //
                // Avoid moving the assumed Pod itself as it's always Unschedulable.
-               // It's intentional to "defer" this operation; otherwise MoveAllToActiveOrBackoffQueue() would
-               // update `q.moveRequest` and thus move the assumed pod to backoffQ anyways.
                if status.IsUnschedulable() {
-                       defer sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, func(pod *v1.Pod) bool {
+                       sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, func(pod *v1.Pod) bool {
                                return assumedPod.UID != pod.UID
                        })
                } else {
+                       // Ensure that the pod goes into the backoff queue.
+                       podInfo.MoveRequestCycle = sched.SchedulingQueue.SchedulingCycle()
                        sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, nil)
                }
        }
```

But setting `podInfo.MoveRequestCycle` here still feels wrong to me. I would prefer an explicit parameter in `FailureHandler`. If such a parameter was added, would `MoveAllToActiveOrBackoffQueue` still be needed here?

I haven't looked at other test failures, but probably the root cause is the same.

pohly commented:
> If such a parameter was added, would MoveAllToActiveOrBackoffQueue still be needed here?

I think it is: not binding the pod to the node after all may make some other pending pods schedulable, so those need to be checked.

Huang-Wei commented:
The context of the "defer" vs. "not defer" was for better handling the event triggering for invalidating a Pod - i.e., un-assume a pod that failed at binding cycle, which is a go routine. So it was implemented to achieve two goals:

- distinguish error and unschedulable.
  - error is treated as un-expected, so the pod itself should be moved
  - unschedulable is treated as an intended binding issue, so the pod itself should not be moved
- `defer` is intentional per the comments said, because in old design, `MoveAllToActiveOrBackoffQueue()` would update `q.moveRequest`.

So with that being said, your proposed [changes](https://github.com/kubernetes/kubernetes/pull/115796#issuecomment-1432730035) did fix the PreBind failure - as that exactly matches the context.

However, in [CI failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115796/pull-kubernetes-integration/1625859602316791808), there are other failures like TestPodTopologySpreadFilter and TestPreemption*, they look tangent, so I'm skeptical those failures are totally different stories. We may need to verify case by case.

pohly commented:
@Huang-Wei: instead of the `podInfo.MoveRequestCycle = sched.SchedulingQueue.SchedulingCycle()` [hack (?)](https://github.com/kubernetes/kubernetes/pull/115796#issuecomment-1432730035) I implemented the (IMHO) cleaner approach with an explicit parameter for `FailureHandler`. The downside of that is that it changes several APIs.

Next I looked into other test failures. They turned out be a variant of the same problem: a cluster event is observed while a pod is being scheduled. Because the queue doesn't know about the pod when the event is handled, the `pInfo.MoveRequestCycle` for the pod does not get bumped and `FailureHandler` doesn't use the backoff queue. The explicit parameter doesn't help in this case because the caller doesn't know that the backoff queue is needed. My solution for this is to track which pods were popped from the queue and are currently being processed - again, an API change :disappointed: 

The explicit parameter is still needed for PreBind failures despite this tracking because PreBind happens after what's currently considered "processing the pod". A more complex tracking of that in `scheduleOne` probably could make that extra parameter unnecessary. I can try that if you don't like the extra parameter.

Next I need to double-check that performance is still good with this variant of the enhancement.

pohly commented:
The performance with the current complete solution is slightly worse than with the initial incomplete one. I would need to do multiple runs to get a statistically significant comparison, but that's not necessary: the current solution is fine and much better than without this PR.

pohly commented:
/retest

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -1050,10 +1073,11 @@ func NewPodNominator(podLister listersv1.PodLister) framework.PodNominator {
-// scheduling queue
-func MakeNextPodFunc(queue SchedulingQueue) func() *framework.QueuedPodInfo {
-	return func() *framework.QueuedPodInfo {
+// MakeNextPodFunctions returns a function to retrieve the next pod from a given
```
MakeNextPodFuncs returns two functions to ...

pohly replied:
Changed.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -729,13 +742,23 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(podInfoList []*framework.
+	// AddUnschedulableIfNotPresent might get called for them, and in
+	// AddUnschedulableIfNotPresent we need to know whether events where
+	// observed while processing them.
+	for pInfo := range p.activePods {
+		pInfo.MoveRequestCycle = p.schedulingCycle
+	}
```
My _gut_-feeling is that this pointer-based `activePods` map can be  optimized, but it's good enough for the current version. Let's maybe add a todo here?

pohly replied:
TODO added.

What would you use instead?

Huang-Wei replied:
I _was_ thinking about something like a bookmark thing, or a (doubly) linkedlist. I'm feeling that we only need to do a partial update to achieve what we want. But maybe an overkill 🤷

Huang-Wei commented on pkg/scheduler/schedule_one.go:
```diff
@@ -62,6 +62,8 @@ const (
 // scheduleOne does the entire scheduling workflow for a single pod. It is serialized on the scheduling algorithm's host fitting.
 func (sched *Scheduler) scheduleOne(ctx context.Context) {
 	podInfo := sched.NextPod()
+	defer sched.DonePod(podInfo)
```
(Continue the comment of keep signature of FailureHandlerFn and AddUnschedulableIfNotPresent here...)

So it makes more sense to ensure the in-flight Pod is tracked for both schedulingCycle AND bindingCycle: by the end the cycle's completion. And thus, we manage the in-flight Pod in a general way, and so maybe we don't need the special logic to deal with pod's invalidation in PreBind? Could you please verify if it's doable? If so, the delta changes would be pretty minimum and decent.

pohly replied:
I was worried that the logic around when to call `Done` would be worse, but it's not that bad: I decided to still have a defer at the top of `scheduleOne`, except that it now uses a variable that gets cleared once the goroutine takes over.

However, there is a snag: once it is time to modify the pod, `schedulingCycle` creates a deep copy of the entire `PodInfo` struct. Then `MoveAllToActiveOrBackoffQueue` updates `MoveRequestCycle` in the original instance whereas `FailureHandler` is called with the clone and therefore doesn't use the backoff queue.

To get around this, I hacked  `schedulingCycle` to do the deep copy only for the pod. When the same `PodInfo` instance is used for the pod, it works - or at least none of the tests fail. My worry is that this will cause problems in other scenarios, like for example when receiving pod update events at exactly the wrong time?

I've pushed a revert of the additional parameter commit, a simple fixup for the review nits and the `Done` change.

pohly replied:
An alternative that keeps the `assumedPodInfo := podInfo.DeepCopy()` in `schedulingCycle` would be to turn `inFlighPods` into a `map[types.UID]int` which maps `UID->MoveRequestCycle.` That's sufficient for the current purpose. But it sounds like you wanted to have more information about in flight pods?

pohly replied:
I've pushed one additional commit with that solution. This feels cleaner to me than the previous hack with not cloning PodInfo.

Huang-Wei replied:
I'm in favor of the latest version (`map[types.UID]cycleNumber`). Thanks!

pohly replied:
With the current `Done` parameter (= pointer) or shall I switch to a UID string?

pohly replied:
Never mind, I hadn't refreshed the GitHub page. I'll switch to UID string on Monday.

Huang-Wei reviewed (commented):
Thanks @pohly . The current version looks reasonable to me, but I'm trying to see if we can make it better :) See the comments below.

chendave commented:
> A benchmark for scheduling pods with resource claim templates finishes in 60 seconds instead of 170s.

this is amazing to look, is there any benchmark test cover this already or need to create a new one?

pohly commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -1050,10 +1073,11 @@ func NewPodNominator(podLister listersv1.PodLister) framework.PodNominator {
+// MakeNextPodFunctions returns a function to retrieve the next pod from a given
+// scheduling queue. For each pod info retrieved via it, the done function
+// must be called when no longer processing the pod.
+func MakeNextPodFunctions(queue SchedulingQueue) (next func() *framework.QueuedPodInfo, done func(*framework.QueuedPodInfo)) {
```
Changed.

pohly commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -157,6 +161,10 @@ type PriorityQueue struct {
+	// activePods holds all pod infos which have been returned by Pop for which Done
+	// hasn't been called yet - in other words, all pods that are currently being
+	// processed.
+	activePods sets.Set[*framework.QueuedPodInfo]
```
Makes sense, changed.

pohly commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -157,6 +161,10 @@ type PriorityQueue struct {
 	lock sync.RWMutex
 	cond sync.Cond
 
+	// activePods holds all pod infos which have been returned by Pop for which Done
```
Changed.

pohly commented:
The DRA benchmark test cases from https://github.com/kubernetes/kubernetes/compare/master...pohly:kubernetes:dra-integration-tests benefit from this. I'm currently trying to get some per-requisites for that merged before creating a PR with them.

pohly commented on pkg/scheduler/schedule_one.go:
```diff
@@ -99,9 +111,15 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
+	// This goroutine becomes responsible for calling DonePod. assumedPodInfo returns a different
+	// pointer, so we have to remember the original one here.
+	donePodInfo := podInfo
+	podInfo = nil
```
The exact pointer isn't needed anymore with the final approach. Should DonePod aka PriorityQueue.Done perhaps be changed to taking a UID as parameter?

Huang-Wei replied:
> Should DonePod aka PriorityQueue.Done perhaps be changed to taking a UID as parameter?

Yes go ahead, that sounds cleaner.

pohly replied:
Done.

Huang-Wei commented on pkg/scheduler/schedule_one.go:
```diff
@@ -62,7 +62,17 @@ const (
+		//
+		// The if check is not strictly needed (DonePod accepts nil),
+		// but checking here is a bit faster.
+		if podInfo != nil {
```
I see how it works. However, I may prefer to pursue the semantics more _explicit_ so more readable. Like, add a variable in the beginning:

```go
breakFromSchedulingCycle := true
```

Then, before starting the bindingCycle, move this variable to true:

```go
breakFromSchedulingCycle = false
```

So here we can change the check to

```go
if breakFromSchedulingCycle {
```

pohly replied:
Done.

Huang-Wei commented on pkg/scheduler/schedule_one.go:
```diff
@@ -120,6 +136,7 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 var clearNominatedNode = &framework.NominatingInfo{NominatingMode: framework.ModeOverride, NominatedNodeName: ""}
 
 // schedulingCycle tries to schedule a single Pod.
+//
```
nit: remove this.

pohly replied:
Done.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -480,11 +486,17 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(pInfo *framework.QueuedPodI
 	pInfo.Timestamp = p.clock.Now()
 
 	// If a move request has been received, move it to the BackoffQ, otherwise move
-	// it to unschedulablePods.
+	// it to unschedulablePods. While the pod was in flight, a more recent cycle
+	// number might have been recorded for it in p.inFlightPods, so here we need
+	// to check both and use the higher value.
```
Let's move this section above L495.

pohly replied:
Done.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -480,11 +486,17 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(pInfo *framework.QueuedPodI
 	// If a move request has been received, move it to the BackoffQ, otherwise move
-	// it to unschedulablePods.
+	// it to unschedulablePods. While the pod was in flight, a more recent cycle
+	// number might have been recorded for it in p.inFlightPods, so here we need
+	// to check both and use the higher value.
```
```suggestion
	// number might have been recorded for it in p.inFlightPods, so in that case
	// we need to honor the more recent cycle number.
```

pohly replied:
Done.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -729,13 +748,23 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(podInfoList []*framework.
+	// AddUnschedulableIfNotPresent might get called for them, and in
+	// AddUnschedulableIfNotPresent we need to know whether events where
+	// observed while processing them.
+	for uid := range p.inFlightPods {
+		p.inFlightPods[uid] = p.schedulingCycle
+	}
```
Can this be moved inside `if activated`?

pohly replied:
No. If there's a single pod in flight and some cluster event comes in that affects the pod, that pod must go into the backoff queue if pod scheduling fails. But `activated` would be false in that case because there was no other pod that got activated.

pohly replied:
While squashing, I was reminded that *not* forcing in-flight pods into the backoff queue unnecessarily was originally the goal for this PR. As seen here, the PR no longer achieves that goal when setting `p.inFlightPods[uid]` for all in-flight pods.

But I [was seeing a benefit](https://github.com/kubernetes/kubernetes/pull/115796#issuecomment-1442134651) from this PR in my integration test branch for DRA. I'm now puzzled why exactly that is, so I double-checked my benchmarking. I think what had happened was that I botched a rebase and ran the benchmark while it didn't set the move request cycle for pods in-flight - that gives good results, but isn't correct :cry: 

The good news is that this PR here can be fixed, it just needs another improvement first that I also have in my full branch: letting plugins register a filter function together with the cluster events that they are interested in. Then in this code here, an unrelated cluster event will not affect an in-flight pod.

I suggest that we change the order of PRs:
- first some general scheduler_perf enhancements (new PR)
- adding DRA testcases to it without any optimization (beware, this will increase overall runtime!)
- cluster event filter callbacks (new PR)
- this PR

Then in each PR which aims at enhancing performance, the impact can be measured with "benchstat" before merging.

/hold

Huang-Wei replied:
> letting plugins register a filter function together with the cluster events that they are interested in.

That sounds like a non-trivial change :) but as long as each PR is well organized and we can tell the perf diff, we should be confident enough to evaluate it.

Huang-Wei reviewed (commented):
The latest version looks pretty neat! Some nits below.

Huang-Wei commented:
cc @kubernetes/sig-scheduling-misc for awareness, this PR can be a thing to improve the performance of scheduler's queuing/re-queuing.

ahg-g commented:
/assign

I would like to take a look at this one.

bart0sh commented:
/triage accepted
/priority important-longterm
/assign @Huang-Wei

Huang-Wei commented:
This will be reviewed after we get to a consensus on #117561.

sanposhiho commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -745,13 +781,71 @@ func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
+				p.unschedulablePods.delete(pod, pInfo.Gated)
+			}
+		} else {
+			gated := pInfo.Gated
```
Is it OK to ignore backoff this case? 
I think the scheduling queue wants to honor backoff time in any cases. 

What if plugin returns `PodImmediatelySchedulable` to many Pods and many Pods goes to activeQ without backoff? Doesn't that affect badly to other Pod's scheduling?

pohly committed:
squash! scheduler: more precise handling of cluster events

pohly committed:
dra: logging updates

klog.Format is new and better than the custom prettyPrint because it generates
different output for JSON and text. In text output, the objects are more
readable because they use multiple lines.

Dumping all potential nodes as structs generated too much output.

pohly committed:
fixup! scheduler/dra: reduce pod scheduling latency

Use PodScheduleImmediately less often. Detect the case when the only
PodSchedulingContext change is the modified SelectedNode, which can be ignored.

pohly committed:
fixup! scheduler: support events with custom "is schedulable" checks - review comments

pohly committed:
fixup! scheduler/dra: reduce pod scheduling latency - review comments

pohly committed:
fixup! scheduler: support events with custom "is schedulable" checks

pohly committed:
fixup! scheduler: support events with custom "is schedulable" checks - reorder checks before moving pod

pohly committed:
fixup! squash! scheduler: more precise handling of cluster events - fix unit test failure

pohly commented:
/close

Replaced by the commits in https://github.com/kubernetes/kubernetes/pull/117844
